web/src/app/ui/kmwuitoolbar.ts

```diff
@@ -23,1200 +63,1363 @@ if(!window['keyman']['ui']['name']) {
+       *
+       * @param       {Object}  a
+       * @param       {Object}  b
+       * @return      {number}
+       **/
+      readonly sortKeyboards = function(a: KeyboardDetail, b: KeyboardDetail) {
+        if(a.RegionCode < b.RegionCode) {
+          return -2;
+        }
+
+        if(a.RegionCode > b.RegionCode) {
+          return 2;
+        }
+
+        if(a.LanguageName < b.LanguageName) {
+          return -1;
+        }
+
+        if(a.LanguageName > b.LanguageName) {
+          return 1;
+        }
 
-      // When a keyboard is activated before init is complete,
-      // the toolbar needs to refresh the loaded keyboard
-      if(ui.lastSelectedKeyboard) {
-        ui.changeKeyboardEvent(ui.lastSelectedKeyboard);
+        return 0;
       }
 
-      // Restore focus
-      keymanweb['focusLastActiveElement']();
-    }
+      /* KeymanWeb Interfaces */
+
+      /**
+       * Function     findListedKeyboard
+       * Scope        Private
+       * @param       {Object}  lang
+       * @return      {?number}
+       * Description  Test if a keyboard is still shown on the toolbar.  Returns index of keyboard in listedKeyboards or null if not found
+       */
+      findListedKeyboard(lang: LanguageEntry | string) {
+        if(typeof lang != 'string') {
+          lang = lang.id;
+        }
+        for(let i = 0; i < this.listedKeyboards.length; i++) {
+          if(this.listedKeyboards[i].lang.id == lang) {
+            return i;
+          }
+        }
 
-    /**
-     * Sort keyboards array returned from keymanweb by region and language
-     *
-     * @param       {Object}  a
-     * @param       {Object}  b
-     * @return      {number}
-     **/
-    ui.sortKeyboards = function(a,b)
-    {
-      if(a['RegionCode'] < b['RegionCode']) return -2;
-      if(a['RegionCode'] > b['RegionCode']) return 2;
-      if(a['LanguageName'] < b['LanguageName']) return -1;
-      if(a['LanguageName'] > b['LanguageName']) return 1;
-      return 0;
-    }
+        return null;
+      }
 
-    /* KeymanWeb Interfaces */
-
-    /**
-     * Function     findListedKeyboard
-     * Scope        Private
-     * @param       {Object}  lang
-     * @return      {?number}
-     * Description  Test if a keyboard is still shown on the toolbar.  Returns index of keyboard in listedKeyboards or null if not found
-     */
-    ui.findListedKeyboard = function(lang)
-    {
-      for(var i = 0; i < ui.listedKeyboards.length; i++)
-        if(ui.listedKeyboards[i].lang.id == lang.id) return i;
-      return null;
-    }
+      /**
+       * Add a keyboard to the list of keyboards available for a language
+       *
+       * @param       {Object}  lang
+       * @param       {Object}  kbd
+       **/
+      addKeyboardToList(lang: LanguageEntry, kbd: KeyboardDetail) {
+        let found = this.findListedKeyboard(lang);
+        if(found == null) {
+          // Add the button
+          if(this.listedKeyboards.length >= this.maxListedKeyboards) {
```

Comment:
And yes, it _is_ enforced.  That means that this method's cache-emulation is heavily-overengineered compared to what's needed for the functionality that is actually offered.